uuid.go

```diff
@@ -344,6 +345,30 @@ func (u *NullUUID) Scan(src interface{}) error {
 	return u.UUID.Scan(src)
 }
 
+// MarshalJSON marshalls the NullUUID as nil or the nested UUID
+func (u NullUUID) MarshalJSON() ([]byte, error) {
+	if u.Valid == false {
+		return json.Marshal(nil)
+	}
+	return json.Marshal(u.UUID)
+}
+
+// UnmarshalJSON unmarshalls a NullUUID
+func (u *NullUUID) UnmarshalJSON(b []byte) error {
+	var s *UUID
+	if err := json.Unmarshal(b, &s); err != nil {
+		return err
+	}
+
+	if s != nil {
+		u.Valid = true
+		u.UUID = *s
+	} else {
+		u.Valid = false
+	}
+	return nil
+}
```

Comment:
@blastrock JSON decoder tokenizes input and it will pass `"null"` to `UnmarshalJSON` in any way.

If you look closer besides stylistic changes I removed intermediate UUID allocation by passing `&u.UUID` to `json.Unmarshal`.
Simple benchmark of original `UnmarshalJSON` and patched one:
```
BenchmarkUnmarshalJSON-4   	 2000000	       879 ns/op	     312 B/op	       3 allocs/op
```
vs
```
BenchmarkUnmarshalJSON-4   	 2000000	       770 ns/op	     288 B/op	       1 allocs/op
```

Benchmark itself:
```
func BenchmarkUnmarshalJSON(b *testing.B) {
    bytes := []byte("\"886313e1-3b8a-5372-9b90-0c9aee199e5d\"")
    u := NullUUID{}
    for i := 0; i < b.N; i++ {
        u.UnmarshalJSON(bytes)
    }
}
```

Just for comparison `UnmarshalText` benchmark does 0 allocations and 10x faster:
```
BenchmarkUnmarshalText-4   	20000000	        88.9 ns/op	       0 B/op	       0 allocs/op
```

I agree with @shurcooL's comment about initializing all fields but it might be `Scan` method does it the most correct way by initializing `UUID` struct member to be `Nil` too.